### PR TITLE
Gracefuly check connection status with server PINGs

### PR DESCRIPTION
**Pull request by LiraNuna**
Currently, every 5 minutes of no data received, the connection will terminate, assuming the server is dead or another issue has occurred. Some servers, such as irc.mzima.net have a higher default PING timeout than 5 minutes, causing false-positives of dead-sockets.

This PR changes this behaviour to instead, PING the server to check if it's still there - when we don't hear back, we THEN terminate the connection.

Fixes https://github.com/Shizmob/pydle/issues/51

### CODE REVIEW

File: pydle/client.py [hunk 1]
```diff
@@ -365,7 +345,17 @@ def _parse_message(self):
     async def handle_forever(self):
         """ Handle data forever. """
         while self.connected:
-            data = await self.connection.recv()
+            try:
+                data = await self.connection.recv(timeout=self.READ_TIMEOUT)
+            except asyncio.TimeoutError:
+                self.logger.debug('>> Receive timeout reached, sending ping to check connection state...')
```
Comment (theunkn0wn1) [inline pydle/client.py]: I think this should be emitted as a warning, as a read timeout is something that should warrant investigation
```suggestion
                self.logger.warning('>> Receive timeout reached, sending ping to check connection state...')
```

File: pydle/connection.py [hunk 2]
```diff
@@ -127,5 +127,5 @@ def stop(self):
         self.writer.write(data)
         await self.writer.drain()
 
-    async def recv(self):
-        return await self.reader.readline()
+    async def recv(self, *, timeout=None):
```
Comment (theunkn0wn1) [inline pydle/connection.py]: If we are going to make `timeout` a keyword only argument by swallowing remaining `*args`, we should raise a warning of any args are present as their presence indicates a usage error.

**Comment by Shizmob** [comment]
> 
> In pydle/connection.py:
> 
> > @@ -127,5 +127,5 @@ def stop(self):
>          self.writer.write(data)
>          await self.writer.drain()
>  
> -    async def recv(self):
> -        return await self.reader.readline()
> +    async def recv(self, *, timeout=None):
> 
> If we are going to make timeout a keyword only argument by swallowing remaining *args, we should raise a warning of any args are present as their presence indicates a usage error.

That is exactly what the lone `*` does:

```python
>>> def foo(*, kwonly=42):
...     pass
...
>>> foo(4)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: foo() takes 0 positional arguments but 1 was given
```


The MR looks otherwise good to me!
Thanks a lot for doing this work, and participating in the discussion upfront about it.
The traditional behaviour has been a long-standing issue caused by my early interpretation
of RFC1459, but I think this new behaviour makes more sense.

File: pydle/connection.py (hunk 2 above)
Comment (Shizmob) [reply pydle/connection.py]: That is exactly what the lone `*` does:

```python
>>> def foo(*, kwonly=42):
...     pass
...
>>> foo(4)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: foo() takes 0 positional arguments but 1 was given
```

File: pydle/client.py (hunk 1 above)
Comment (LiraNuna) [reply pydle/client.py]: Fair suggestion!

**Comment by Shizmob** [comment]
The MR looks otherwise good to me!

Thanks a lot for doing this work, and participating in the discussion upfront about it.
The traditional behaviour has been a long-standing issue caused by my early interpretation
of RFC1459, but I think this new behaviour makes more sense.

**Comment by LiraNuna** [comment]
> master is reserved for releases, whereas develop is used as the destination for pull requests for the next release.

I apologize for that. I believe there's a way to select a default PR branch in github for that purpose.

**Comment by Shizmob** [comment]
> I apologize for that. I believe there's a way to select a default PR branch in github for that purpose.

Good call, I've changed this in the project settings now. This wasn't done before as I believe @theunkn0wn1 has no access to those settings, and moving the main development HEAD to `develop` was mostly his work.

File: pydle/connection.py (hunk 2 above)
Comment (theunkn0wn1) [reply pydle/connection.py]: TIL. I wasn't aware of that language feature.

**Comment by LiraNuna** [comment]
Please note I have changed `PING_TIMEOUT` to `READ_TIMEOUT`, which MAY be a breaking change to some users. I would probably add a note in the CHANGELOG before releasing or add a compatibility layer of some sort.

**Comment by theunkn0wn1** [comment]
Yes I had noticed that, Upon consideration we should keep the old name, or alias it with a property. Otherwise it will take a major point release to merge this.

**Comment by Shizmob** [comment]
I did notice that -- I was going to propose to add a property that proxies to `READ_TIMEOUT` for compatibility, but figured that didn't need to impact the MR itself.

Unrelatedly, I think the current reconnect code is a bit flaky as well; once that has been fixed, it seems the connection problems would finally be gone for good.

**Comment by theunkn0wn1** [comment]
I was going to touch the reconnect code soon anyways, as i intend to work in #117 over the weekend. My current idea will require changing how pydle expects a disconnect.

**Comment by LiraNuna** [comment]
> I did notice that -- I was going to propose to add a property that proxies to READ_TIMEOUT for compatibility, but figured that didn't need to impact the MR itself.

Just so we're on the same page, is there anything you'd like me to do? should I add the proxy property or defer to a later PR? I would love to see this merged and ideally released because right now I am using this library with a modified `sites-packages` for my issue to be resolved.

Personally I don't think this is a huge breaking change because if someone DID modify `PING_TIMEOUT`, it's most likely to increase the amount in order to avoid the incorrect behavior, now that it's gone it's really not needed anymore so it won't really break anything behavior wise, as the timeout handling is now in the library instead of being a client's responsibility.

Let me know how you'd like to proceed.

**Comment by theunkn0wn1** [comment]
Since this resolves an immediate issue here is what I think is best:
We merge this PR now as-is, with the understanding that a proxy property will be added in a later PR.

I am drafting this fix up shortly, so we can do a point release soon. If anyone is both dependent on modifying the renamed variable and is using the develop branch, they should be aware of the risk to using a development branch rather than a release branch.

**Pull request merged**